### PR TITLE
refactor: update evaluate_dataset to take in a dataset instead of dataset config

### PR DESCRIPTION
**Pull request by danielezhu**
*Description of changes:*
This PR refactors the `evaluate_dataset` method to consume a dataset directly, instead of a dataset config. This will allow `evaluate_dataset` to be compatible with more eval algorithms.

The `verify_model_determinism` function has been updated in preparation for its usage in Summarization Accuracy Semantic Robustness. By taking in a prompt template and model input column, we can verify model determinism prior to executing the transforms for prompt-generation and model-invocation. This will allow SASR's `evaluate` method to follow the same overall template as all other algos.

This PR additionally removes the `BertscoreHelperModel` class, as we now use `BertscoreModel`.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/fmeval/eval_algorithms/util.py [hunk 1]
```diff
@@ -427,40 +415,58 @@ def create_model_invocation_pipeline(model: ModelRunner, prompt_template: str) -
...
+    if prompt_template:
+        util.require(model, "A model must be provided as well if the provided prompt template is not None.")
+    if model:
+        prompt_template = get_default_prompt_template(dataset_name) if not prompt_template else prompt_template
+        model_invocation_pipeline = create_model_invocation_pipeline(model, prompt_template)
+        pipeline = TransformPipeline([model_invocation_pipeline, pipeline])
+    else:
+        try:
+            validate_dataset(dataset, [DatasetColumns.MODEL_OUTPUT.value.name])
+        except EvalAlgorithmClientError:
```
Comment (danielezhu) [inline src/fmeval/eval_algorithms/util.py:463]: The try-except is just for providing a more specific error message than the generic one provided by `validate_dataset`

File: src/fmeval/eval_algorithms/util.py [hunk 2]
```diff
@@ -427,40 +415,58 @@ def create_model_invocation_pipeline(model: ModelRunner, prompt_template: str) -
...
     :param agg_method: The aggregation method to use when aggregating the computed metric values.
         Currently, only MEAN is supported.
     :param save: If set to true, prompt responses and scores will be saved to a file.
         The path that this file is stored at is configured by `eval_results_path`.
 
-    :return: A list of EvalOutput objects.
+    :return: An EvalOutput object encapsulating the results of the evaluation.
     """
+    if prompt_template:
+        util.require(model, "A model must be provided as well if the provided prompt template is not None.")
```
Comment (malhotra18) [inline src/fmeval/eval_algorithms/util.py:455]: why do we need an exception for this? Why not just log a warning? If customer does provide prompt_template along with ModelOutput, we are simply ignoring the prompt template.

File: src/fmeval/eval_algorithms/summarization_accuracy_semantic_robustness.py [hunk 3]
```diff
@@ -56,6 +55,7 @@
     generate_mean_delta_score,
     generate_model_predict_response_for_dataset,
 )
+from fmeval.helper_models import BertscoreModelTypes
```
Comment (malhotra18) [inline src/fmeval/eval_algorithms/summarization_accuracy_semantic_robustness.py:58]: Apologies if I missed it earlier.

We should be sticking to the terminology of HelperModels. This is gonna be useful in future too when we introduce other helper models.

**Review by malhotra18 — CHANGES_REQUESTED** [review]

File: src/fmeval/eval_algorithms/general_semantic_robustness.py [hunk 4]
```diff
@@ -330,18 +330,17 @@ def evaluate(
             dataset_prompt_template = (
                 get_default_prompt_template(dataset_config.dataset_name) if not prompt_template else prompt_template
             )
-            model_invocation_pipeline = create_model_invocation_pipeline(model, dataset_prompt_template)
-            dataset = model_invocation_pipeline.execute(dataset)
-            is_deterministic = verify_model_determinism(model, dataset, DatasetColumns.PROMPT.value.name)
-            pipeline = self._build_pipeline(model, dataset_prompt_template, is_deterministic=is_deterministic)
-            eval_output = compute_and_aggregate_metrics(
-                pipeline=pipeline,
+            is_deterministic = verify_model_determinism(model, dataset, dataset_prompt_template)
```
Comment (malhotra18) [inline src/fmeval/eval_algorithms/general_semantic_robustness.py:333]: Note: We need to do a model check before this for semantic robustness algorithms.

**Review by malhotra18 — CHANGES_REQUESTED** [review]

File: src/fmeval/eval_algorithms/general_semantic_robustness.py (hunk 4 above)
Comment (danielezhu) [reply src/fmeval/eval_algorithms/general_semantic_robustness.py:333]: Since the `model` argument is not `Optional`, I will omit the explicit check. However, I have updated the docstring to indicate the reasoning for why the model is required. The same docstring update has been applied to the other SR algos.

**Review by oyangz — APPROVED** [review]

File: src/fmeval/eval_algorithms/helper_models/helper_model.py [hunk 5]
```diff
@@ -140,14 +147,10 @@ def get_helper_scores(self, text_input: List[str]) -> Dict[str, List[float]]:  #
         """
         inputs = self._tokenizer(text_input, return_tensors="pt", truncation=True, padding=True).to(self._model.device)
         scores = torch.sigmoid(self._model(**inputs)[0]).cpu().detach().numpy()
-        results = {}
```
Comment (danielezhu) [inline src/fmeval/eval_algorithms/helper_models/helper_model.py:146]: Since `text_input` is always a `List` and not a `str`,  we don't need the `isinstance(text_input, str)` section. The fact that it is always a `List` is indicated in the type annotation of the function, and  everywhere that `DetoxifyHelperModel.get_helper_scores` is called, we pass a list: see `DetoxifyHelperModel.__call__` and `evaluate_sample` for `toxicity.py`).

Comment (danielezhu) [reply src/fmeval/eval_algorithms/helper_models/helper_model.py:146]: Not to mention that if we use the old code, then the output type annotation of this function is wrong; the returned type will be a `Dict[str, float]`. I have already manually verified this.

**Review by malhotra18 — APPROVED** [review]
